src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClientX/Handlers/TransportCreation/TcpTransportCreationHandler.cs

```diff
@@ -0,0 +1,198 @@
+using System.Collections.Generic;
+using System.IO;
+using System.Net;
+using System.Net.Sockets;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Data.SqlClient;
+using Microsoft.Data.SqlClient.SNI;
+
+namespace Microsoft.Data.SqlClientX.Handlers.TransportCreation
+{
+    /// <summary>
+    /// Handler for connecting via TCP.
+    /// </summary>
+    internal sealed class TcpTransportCreationHandler : ReturningHandler<ConnectionHandlerContext, Stream>
+    {
+        private const int KeepAliveIntervalSeconds = 1;
+        private const int KeepAliveTimeSeconds = 30;
+
+        #if NET8_0_OR_GREATER
+        private static readonly TimeSpan DefaultPollTimeout = TimeSpan.FromSeconds(30);
+        #else
+        private const int DefaultPollTimeout = 30 * 100000; // 30 seconds as microseconds
+        #endif
+
+        /// <inheritdoc />
+        public override async ValueTask<Stream> Handle(ConnectionHandlerContext parameters, bool isAsync, CancellationToken ct)
+        {
+            // This handler cannot process if the protocol does not contain TCP
+            if (parameters.DataSource.ResolvedProtocol is not (DataSource.Protocol.Admin or DataSource.Protocol.TCP or DataSource.Protocol.None))
+            {
+                return await HandleNext(parameters, isAsync, ct);
+            }
+
+            ct.ThrowIfCancellationRequested();
+
+            // DNS lookup
+            IPAddress[] ipAddresses = isAsync
+                ? await Dns.GetHostAddressesAsync(parameters.DataSource.ServerName, ct).ConfigureAwait(false)
+                : Dns.GetHostAddresses(parameters.DataSource.ServerName);
+            if (ipAddresses is null || ipAddresses.Length == 0)
+            {
+                throw new SocketException((int)SocketError.HostNotFound);
+            }
+
+            // If there is an IP version preference, apply it
+            switch (parameters.ConnectionString.IPAddressPreference)
+            {
+                case SqlConnectionIPAddressPreference.IPv4First:
+                    Array.Sort(ipAddresses, IpAddressVersionComparer.InstanceV4);
+                    break;
+
+                case SqlConnectionIPAddressPreference.IPv6First:
+                    Array.Sort(ipAddresses, IpAddressVersionComparer.InstanceV6);
+                    break;
+
+                case SqlConnectionIPAddressPreference.UsePlatformDefault:
+                default:
+                    // Not sorting necessary
+                    break;
+            }
+
+            // Attempt to connect to one of the matching IP addresses
+            // @TODO: Handle opening in parallel
+            Socket socket = null;
+            var socketOpenExceptions = new List<Exception>();
+
+            int portToUse = parameters.DataSource.ResolvedPort < 0
+                ? parameters.DataSource.Port
+                : parameters.DataSource.ResolvedPort;
+            var ipEndpoint = new IPEndPoint(IPAddress.None, portToUse); // Allocate once
```

Comment:
if you want to be consistent with the capitalization in the class name and the parameter name on line 111, you can rename the local variable to `ipEndPoint`

Reply:
Good catch. I'm going with the precedent dotnet set with `IPEndPoint` (but fixing the acronym capitalization).